### PR TITLE
Fix semantic changes related to numeric literals

### LINKED ISSUE #33316: Type of floating-point-literal when contextually expected type is int|decimal
Please refer to the discussion of https://github.com/ballerina-platform/ballerina-spec/issues/943.

### LINKED ISSUE #33318: an out of range float  should be a compile-time error
Please refer to discussion https://github.com/ballerina-platform/ballerina-spec/issues/939

### PR DESCRIPTION
**Pull request by chiranSachintha**
Fixes #33316 
Fixes #33318 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by hasithaa** [comment]
@chiranSachintha Please send this PR to stage-slbeta4 branch.

**Comment by MaryamZi** [comment]
> @chiranSachintha Please send this PR to stage-slbeta4 branch.

I've changed the base back to master temporarily so that the PR can be reviewed until @chiranSachintha filters out just the relevant changes for stage-slbeta4.

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibSubTypeTest.java [hunk 1]
```diff
@@ -206,7 +206,6 @@ public void testNegativeIntSubType() {
         BAssertUtil.validateError(result, err++, "incompatible types: expected 'decimal'," + FOUND_UNSIGNED_32, 237,
                                   17);
 
-        BAssertUtil.validateError(result, err++, EXPECT_BYTE + FOUND_INT, 251, 14);
```
Comment (MaryamZi) [inline langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibSubTypeTest.java:209]: `byte r = 1 | 1;` also seems to be allowed now. Didn't we discuss to raise this in the spec?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/bytetype/BByteValueNegativeTest.java [hunk 2]
```diff
@@ -81,20 +81,15 @@ public void testByteValueNegative() {
     @Test(description = "Test byte shift operators negative", groups = { "disableOnOldParser" })
     public void invalidByteShiftOperators() {
         CompileResult result = BCompileUtil.compile("test-src/types/byte/byte-shift-operators-negative.bal");
-        Assert.assertEquals(result.getErrorCount(), 13);
+        Assert.assertEquals(result.getErrorCount(), 8);
         int i = 0;
-        BAssertUtil.validateError(result, i++, "incompatible types: expected 'byte', found 'int'", 2, 14);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/bytetype/BByteValueNegativeTest.java:86]: Same comment as above.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/decimaltype/BDecimalValueNegativeTest.java [hunk 3]
```diff
@@ -33,15 +33,13 @@
...
         CompileResult compileResult = BCompileUtil.compile("test-src/types/decimal/decimal_value_negative.bal");
-        Assert.assertEquals(compileResult.getErrorCount(), 19);
+        Assert.assertEquals(compileResult.getErrorCount(), 18);
         int index = 0;
         BAssertUtil.validateError(compileResult, index++, "missing semicolon token", 21, 1);
         BAssertUtil.validateError(compileResult, index++, "missing binary operator", 23, 21);
         BAssertUtil.validateError(compileResult, index++, "undefined symbol 'g'", 23, 21);
         BAssertUtil.validateError(compileResult, index++, "leading zeros in numeric literals", 26, 17);
         BAssertUtil.validateError(compileResult, index++, "Hexadecimal '-0x' too small", 29, 17);
-        BAssertUtil.validateError(compileResult, index++, "incompatible types: expected 'decimal', found 'int'", 29,
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/decimaltype/BDecimalValueNegativeTest.java:43]: This is for `decimal e = -0xX1231.12P1;`. 

https://ballerina.io/ballerina-spec/spec.html#section_5.2.4.2 says 

> Any decimal value can be represented by a DecimalFloatingPointNumber with an optional DecimalTypeSuffix.

So is it correct for this error to no longer be logged?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/decimaltype/BDecimalValueNegativeTest.java [hunk 4]
```diff
@@ -60,14 +58,10 @@ public void testDecimalValue() {
...
         CompileResult negative = BCompileUtil.compile("test-src/types/decimal/decimal_value_negative_literal.bal");
-        Assert.assertEquals(negative.getErrorCount(), 17);
+        Assert.assertEquals(negative.getErrorCount(), 13);
         int i = 0;
         BAssertUtil.validateError(negative, i++, "incompatible types: expected 'decimal', found 'float'", 20, 17);
         BAssertUtil.validateError(negative, i++, "incompatible types: expected 'decimal', found 'float'", 21, 17);
-        BAssertUtil.validateError(negative, i++, "incompatible types: expected 'decimal', found 'int'", 22, 17);
-        BAssertUtil.validateError(negative, i++, "incompatible types: expected 'decimal', found 'int'", 23, 17);
-        BAssertUtil.validateError(negative, i++, "incompatible types: expected 'decimal', found 'int'", 24, 17);
-        BAssertUtil.validateError(negative, i++, "incompatible types: expected 'decimal', found 'int'", 25, 17);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/decimaltype/BDecimalValueNegativeTest.java:70]: Same comment as above.

File: tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-var-type.bal [hunk 5]
```diff
@@ -390,7 +390,7 @@ function testQueryConstructingTableWithVar() returns error? {
                   where user.age > 21 && user.age < 60
                   select {user};
 
-    assertEquality(true, result1 is table<record {| User user; |}> key(user));
+    assertEquality(true, result1 is table<record {| User user; |}>);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-var-type.bal:393]: Why are these changes needed?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/float/float-value-negative.bal [hunk 6]
```diff
@@ -3,3 +3,11 @@ function testInvlaidFloatValue() returns (float) {
...
     return b;
 }
+
+function testInvlaidFloatValue2() returns (float) {
+    float b = 999e9999999999;
+    float c = 999e-9999999999;
+    decimal|float d = 999e9999999999;
+    int|decimal|float e = 99.9E99999999;
+    int|decimal|float f = 99.9E-99999999;
+}
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/float/float-value-negative.bal:13]:
```suggestion
}

```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java [hunk 7]
```diff
@@ -1345,14 +1345,28 @@ public void visit(BLangFiniteTypeNode finiteTypeNode) {
...
         finiteTypeSymbol.type = finiteType;
 
         resultType = finiteType;
     }
 
+public BType getTypeOfNumericLiteral(NodeKind nodeKind, BType type) {
+    if (nodeKind != NodeKind.NUMERIC_LITERAL || type.getKind() != TypeKind.UNION) {
+        return type;
+    }
+    Set<BType> memberTypes = ((BUnionType) type).getMemberTypes();
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java:1360]: Can you share examples for when these are hit?

Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java:1360]: For these kinds of scenarios
```ballerina
public function main() {
    "value" _ = "value";
}

```

File: tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-var-type.bal (hunk 5 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-var-type.bal:393]: To define a key that field should be read-only. right? Due to this change(https://github.com/ballerina-platform/ballerina-lang/blob/21d5a766e8b4592933177c1c252bf84413a3ce89/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L5580) I fixed this test case.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/decimaltype/BDecimalValueNegativeTest.java (hunk 3 above)
Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/decimaltype/BDecimalValueNegativeTest.java:43]: https://ballerina.io/ballerina-spec/spec.html#section_6.5 also mentions this (before we consider the expected type).

> if the numeric-literal is a HexFloatingPointLiteral, then the basic type is float;

**Comment by chiranSachintha** [comment]
As discussed offline create separate PR(https://github.com/ballerina-platform/ballerina-lang/pull/33643) to fix this issue.
